_posts/2022-10-18-torch-linalg/torch-linalg.Rmd

```diff
@@ -0,0 +1,632 @@
+---
+title: "Five ways to do least squares (with torch)"
+description: >
+  Get to know torch's linalg module, all while learning about different ways to do least-squares regression from scratch. This post, again, is a condensed version of the corresponding chapter in the forthcoming book, Deep Learning and Scientific Computing with R torch, to be published by CRC Press. 
+author:
+  - name: Sigrid Keydana
+    affiliation: RStudio
+    affiliation_url: https://www.rstudio.com/
+slug: keydanatorchlinalg
+date: 2022-10-18
+categories:
+  - Torch
+  - R
+  - Concepts
+  - Tabular data
+output:
+  distill::distill_article:
+    self_contained: false
+    toc: true
+preview: images/squirrel.jpg
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE, fig.width = 6, fig.height = 6)
+```
+
+*Note: This post is a condensed version of a chapter from part three of the forthcoming book, Deep Learning and Scientific Computing with R torch. Part three is dedicated to scientific computation beyond deep learning. As throughout the book, but -- if possible -- to an even higher degree, there I focus on the underlying concepts, striving to explain them in as "verbal" a way as I can. This does not mean skipping the equations; it means taking care to explain why they are the way they are.*
+
+How do you compute linear least-squares regression? In R, using `lm()`; in `torch`, there is `linalg_lstsq()`.
+
+Where R, sometimes, hides complexity from the user, high-performance computation frameworks like `torch` tend to ask a bit more up-front effort, be it careful reading of documentation, or playing around some, or both. For example, here is the central piece of documentation for `linalg_lstsq()`, elaborating on the `driver` parameter to the function:
+
+```
+`driver` chooses the LAPACK/MAGMA function that will be used.
+For CPU inputs the valid values are 'gels', 'gelsy', 'gelsd, 'gelss'.
+For CUDA input, the only valid driver is 'gels', which assumes that A is full-rank.
+To choose the best driver on CPU consider:
+  -   If A is well-conditioned (its condition number is not too large), or you do not mind some precision loss:
+     -   For a general matrix: 'gelsy' (QR with pivoting) (default)
+     -   If A is full-rank: 'gels' (QR)
+  -   If A is not well-conditioned:
+     -   'gelsd' (tridiagonal reduction and SVD)
+     -   But if you run into memory issues: 'gelss' (full SVD).
+```
+
+
+Whether you'll need to know this will depend on the problem you're solving. But if you do, it certainly will help to have an idea what is being talked about there, if only in a high-level way.
```

Comment:
```suggestion
Whether you'll need to know this will depend on the problem you're solving. But if you do, it certainly will help to have an idea what is being described there, if only in a high-level way.
```

Reply:
Described, or discussed, or something of the like. For some reason “talked about” always sounds odd to me when referencing something non-verbal in writing.

Reply:
yeah, right! I think "alluded to" will do fine ...